plugin/trino-singlestore/src/main/java/io/trino/plugin/singlestore/SingleStoreClientModule.java

```diff
@@ -63,4 +70,12 @@ public static ConnectionFactory createConnectionFactory(BaseJdbcConfig config, C
                 connectionProperties,
                 credentialProvider);
     }
+
+    @VisibleForTesting
+    static void checkForLegacyDriverConnectionUrl(String connectionUrl)
+    {
+        requireNonNull(connectionUrl, "connectionUrl is null");
+        // including "jdbc:" portion in case-insensitive match in case other url parts also contain "mariadb" literal
+        checkArgument(!connectionUrl.matches("(?i)jdbc:mariadb:.*"), DRIVER_PROTOCOL_ERROR);
```

Comment:
Validation can be done directly in the Config class, using `@javax.validation.constraints.Pattern`

since it's connector specific, you'd need to copy 
https://github.com/trinodb/trino/blob/8d5b360b73d1e91625abd43ed731c8deaed7ece0/plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java#L45-L58 
into a new class

one config property from a file can map to one or more java-level properties, so duplication isn't a problem

Reply:
That's an interesting approach, but I feel like overriding the configuration like that to add annotation-based validation is overly complex (if singlestore already had its own extension of BaseJdbcConfig, or any other Jdbc-based connector overrode the configuration, it might be more correct to follow that approach).
I'd consider adding a "UrlValidationRegex" so JdbcConnectors could easily supply the regex to validate (and possibly error message), but this is an attempt to give a better error message for a (possibly rare) mistake, so modifying the configuration would be overkill.

Reply:
The benefit of having that validation live in configuration is that it doesn't matter how you create the configuration it'll always be validated. We do this in other config classes as well - see PinotConfig, BaseJdbcConfig, BigQueryConfig for examples.

Reply:
The closest analogy I can see to this is `MySqlJdbcConfig` which adds validation methods to the config object.

Implementing it this way involves changing the MySqlClientModule (`.bindConfig(MySqlJdbcConfig.class)`), but in that case the validation isn't overriding a field.

So if I understand correctly, we want something like this:
```
public class SingleStoreJdbcConfig
        extends BaseJdbcConfig
{
    private? final String connectionUrl; // masks field from superclass

 @Pattern(regexp = "^jdbc:mariadb.*", flags = CASE_INSENSITIVE)  // I need to negate this + throw a specific error message... any suggestions?
 public String getConnectionUrl() 
 { 
     return connectionUrl; 
 } 
  
 @Config("connection-url") 
 public BaseJdbcConfig setConnectionUrl(String connectionUrl) 
 { 
     this.connectionUrl = connectionUrl; 
     return this; 
 } 

}

```
? (This is what I'm imagining when I see the suggestion Piotr's suggestion, and I'm saying it's overly complex / not an improvement (and I'm not sure how it would encapsulate the logic of "error in the face of a *specific* misconfiguration, not de-facto validation of a valid URL").

Reply:
From PinotConfig, I see this:
```
    @PostConstruct
    public void validate()
    {
        checkState(
                !countDistinctPushdownEnabled || aggregationPushdownEnabled,
                "Invalid configuration: pinot.aggregation-pushdown.enabled must be enabled if pinot.count-distinct-pushdown.enabled");
    }
```
From MySqlJdbcConfig I see this:
```
public class MySqlJdbcConfig
        extends BaseJdbcConfig
{
    @AssertTrue(message = "Invalid JDBC URL for MySQL connector")
    public boolean isUrlValid() 
    {...}
```
I can combine those 2 approaches to wire the existing check into the configuration - but I can't override `BaseJdbcConfig#private String connectionUrl;` to add validation using `@javax.validation.constraints.Pattern`